### PR TITLE
feat: add hybrid click instrumentation for views and compose

### PR DESCRIPTION
**Pull request by ashishz-vunet**
## Summary
- Add a new `instrumentation/hybrid-click` module to capture hybrid tap interactions across both View and Jetpack Compose surfaces.
- Wire hybrid click into the Android Agent DSL via `hybridClick { ... }`, including configurable `activeContextWindowMillis`.
- Add `ConfigurableHybridClickInstrumentation` so the agent can pass runtime configuration into instrumentation implementations.
- Update public API signature files and docs (`README.md` and module README) to expose and describe the new instrumentation.
- Add tests for key behavior (`HybridClickConfigurationTest`, `HybridLabelResolverTest`) and improve fake loader support for typed instrumentation lookup in tests.

## Test plan
- [x] `./gradlew :android-agent:test :instrumentation:hybrid-click:test`
- [x] Validate in a sample hybrid app that View + Compose click events are emitted with expected labels/targets.
- [ ] Verify `activeContextWindowMillis` tuning changes event grouping behavior as expected.

### CODE REVIEW

**Comment by ashishz-vunet** [comment]
@copilot resolve the merge conflicts in this pull request

File: instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickInstrumentation.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ */
+class HybridClickInstrumentation : AndroidInstrumentation, ConfigurableHybridClickInstrumentation {
+    override val name: String = "hybrid.click"
+    private var activeContextWindowMillis: Long = DEFAULT_ACTIVE_CONTEXT_WINDOW_MILLIS
+
+    /**
+     * Creates the tracer used for hybrid click spans and registers lifecycle callbacks that
+     * attach/detach per-window touch interception as activities resume/pause.
+     */
+    override fun install(ctx: InstallationContext) {
```
Comment (ganeshnk-vunet) [inline instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickInstrumentation.kt]: Blocking: install() registers ActivityLifecycleCallbacks but the callback instance isn’t stored and there’s no uninstall() implementation to unregister. If install happens more than once (tests / multiple initializations), this can duplicate spans and leak callbacks. Please store the callback reference and unregister in uninstall(), and make install idempotent.

File: instrumentation/android-instrumentation/api/android-instrumentation.api [hunk 2]
```diff
@@ -1,6 +1,18 @@
 public abstract interface class io/opentelemetry/android/instrumentation/AndroidInstrumentation {
```
Comment (ganeshnk-vunet) [inline instrumentation/android-instrumentation/api/android-instrumentation.api:1]: This .api file changes AndroidInstrumentation.install()/uninstall() to take InstallationContext. Please ensure the actual AndroidInstrumentation source interface is updated to match, and update all existing instrumentation implementations accordingly. Otherwise we risk compilation issues or publishing an artifact whose API signatures don’t match the shipped code. Also ensure InstallationContext is committed in the correct source set and is referenced consistently.

**Review by ganeshnk-vunet — CHANGES_REQUESTED** [review]

File: instrumentation/android-instrumentation/api/android-instrumentation.api (hunk 2 above)
Comment (ashishz-vunet) [reply instrumentation/android-instrumentation/api/android-instrumentation.api:1]: Thanks for flagging this. I re-verified the API/source alignment after resolving merge conflicts.

- `AndroidInstrumentation` source uses `InstallationContext` for both `install()` and `uninstall()`.
- `instrumentation/android-instrumentation/api/android-instrumentation.api` matches those exact signatures.
- `InstallationContext` is present in the correct source set and referenced consistently.
- I also checked for any leftover old signatures (`Context, OpenTelemetryRum`) and none remain.

So this part should now be consistent and safe for publishing.

**Comment by ashishz-vunet** [comment]
Addressed the requested changes in this PR.

### 1) Hybrid click lifecycle safety (blocking comment)
Implemented in `HybridClickInstrumentation`:
- Store the registered `ActivityLifecycleCallbacks` instance.
- Make `install()` idempotent (no-op when already installed).
- Implement `uninstall()` to unregister callbacks and clear internal state.

This prevents duplicate callback registration and avoids leaked callbacks/spans on repeated initialization.

### 2) API/source alignment for `AndroidInstrumentation`
Re-verified after merge-conflict resolution:
- `AndroidInstrumentation` source uses `InstallationContext` for `install()` and `uninstall()`.
- `instrumentation/android-instrumentation/api/android-instrumentation.api` matches those signatures.
- No lingering old `Context + OpenTelemetryRum` signatures remain.

Validation run:
- `./gradlew :instrumentation:hybrid-click:test :android-agent:test`

File: instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickInstrumentation.kt (hunk 1 above)
Comment (ashishz-vunet) [reply instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickInstrumentation.kt]: Implemented in HybridClickInstrumentation:

Store the registered ActivityLifecycleCallbacks instance.
Make install() idempotent (no-op when already installed).
Implement uninstall() to unregister callbacks and clear internal state.
This prevents duplicate callback registration and avoids leaked callbacks/spans on repeated initialization.

File: instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickEventGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+     * Target resolution prefers Compose views when available and falls back to View-based
+     * detection. The span includes semantic app widget/location attributes and custom
+     * attributes for `view.label` and `view.source`.
+     */
+    fun generateClick(motionEvent: MotionEvent?) {
+        val window = windowRef?.get() ?: return
+        if (motionEvent == null || motionEvent.actionMasked != MotionEvent.ACTION_UP) {
+            return
+        }
+
```
Comment (rakshiths-vunet) [inline instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickEventGenerator.kt:72]: One issue I’m noticing in the current implementation is how `ACTION_UP` is treated as a potential click without distinguishing it from the end of a scroll gesture.

In typical touch interaction flows, both a genuine tap and the end of a scroll produce an `ACTION_UP` event. Since the code only checks for `ACTION_UP`, it ends up processing both scenarios identically. This means that even when a user is simply scrolling (with significant finger movement), the system still attempts to resolve a tap target.

As a result:

* Scroll gestures can trigger unnecessary traversal of the entire view hierarchy
* In cases where no clickable target is found, this work is effectively wasted
* During frequent scrolling (e.g., lists), this can happen repeatedly, leading to avoidable CPU work and memory churn

The core concern is that there’s currently no distinction between a true click (minimal movement) and a scroll/drag interaction, even though both end with `ACTION_UP`. This can lead to misclassification of user intent and unnecessary processing overhead.

File: instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickEventGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import io.opentelemetry.semconv.incubating.AppIncubatingAttributes.APP_SCREEN_COORDINATE_X
+import io.opentelemetry.semconv.incubating.AppIncubatingAttributes.APP_SCREEN_COORDINATE_Y
+import io.opentelemetry.semconv.incubating.AppIncubatingAttributes.APP_WIDGET_ID
+import io.opentelemetry.semconv.incubating.AppIncubatingAttributes.APP_WIDGET_NAME
+import java.lang.ref.WeakReference
+
+internal class HybridClickEventGenerator(
+    private val tracer: Tracer,
+    private val viewTapTargetDetector: HybridViewTapTargetDetector = HybridViewTapTargetDetector(),
+    private val activeContextWindowMillis: Long = DEFAULT_ACTIVE_CONTEXT_WINDOW_MILLIS,
```
Comment (rakshiths-vunet) [inline instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickEventGenerator.kt:26]: One concern with the current approach is the reliance on `activeContextWindowMillis` as a time-based mechanism to correlate `ui.click` spans with downstream async work (e.g., network calls).

This introduces a limitation: correlation is inferred based on *timing*, not *causality*. In scenarios with concurrent or delayed operations (multiple rapid clicks, debounced actions, background retries, etc.), this can lead to incorrect parent-child relationships or missed associations entirely.

For example:

* A network call triggered slightly after the window closes would not be linked to the originating click
* Concurrent API calls from different user actions within overlapping windows may be incorrectly grouped

Would it be possible to explore a context propagation–based approach instead of (or in addition to) the time window?

Specifically:

* Capturing the `Context` at click time (`Context.current().with(span)`)
* Propagating it explicitly through async boundaries (e.g., coroutines, executors, callbacks)
* Allowing downstream instrumentation (e.g., OkHttp) to pick up the correct parent span from the active context

This would enable deterministic parent-child relationships even with concurrency, rather than relying on heuristic timing windows.

Curious to hear thoughts on whether context propagation could be integrated here, or if there are constraints that led to choosing the current time-window model.

**Review by rakshiths-vunet — CHANGES_REQUESTED** [review]

File: instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickEventGenerator.kt (hunk 4 above)
Comment (ashishz-vunet) [reply instrumentation/hybrid-click/src/main/kotlin/io/opentelemetry/android/instrumentation/hybrid/click/HybridClickEventGenerator.kt:26]: Great point — we agree context propagation is the more deterministic long-term direction for modeling click-to-async causality.

For this PR, we intentionally kept a configurable time-window approach (activeContextWindowMillis, default 500ms) as a pragmatic baseline so it works broadly without requiring app-wide propagation across coroutines/executors/callbacks.

We’ll keep this as-is for now, and track context-propagation-based correlation as a follow-up enhancement. That future work would capture click context at source and propagate it across async boundaries so downstream instrumentations (e.g., network) can attach using causality rather than timing heuristics.

**Review by ganeshnk-vunet — APPROVED** [review]

**Review by rakshiths-vunet — APPROVED** [review]

**Comment by gopal-vunet** [comment]
@ashishz-vunet As discussed, Please point this PR to a separate feature/working branch. The current compose click instrumentation captures only the compose container. It is not capturing the nodes(actual compose views) inside the compose container. 

Please create the issues for the following before merging.
- Detecting Compose Nodes
- Currently multiple click spans are generated on multiple clicks - This needs to discussed and designed better along with rage taps.

**Review by gopal-vunet — APPROVED** [review]
Kindle create issues on open items and link to this PR as open items

**Review by gopal-vunet — APPROVED** [review]

**Review by sid-vunet — APPROVED** [review]
Approved
